src/auth/linking.ts

```diff
@@ -0,0 +1,299 @@
+  const { data } = event
+  const message = new TextDecoder().decode(data.arrayBuffer ? await data.arrayBuffer() : data)
+  console.debug("Linking Status", ls)
+  console.debug("message (raw)", message)
+  switch (ls.role) {
+    case "CONSUMER":
+      await handleConsumerMessage(message)
+      break
+    case "PRODUCER":
+      await handleProducerMessage(message)
+      break
+  }
+}
+
+const handleProducerMessage = async (message: string): Promise<void> => {
+  if (ls.step === "BROADCAST") {
+    await sendSessionKey(message)
+  } else if (ls.step === "NEGOTIATION") {
+    const response = await handleUserChallenge(message)
+    if (response) {
+      const json = JSON.parse(response)
+      console.log("PIN to SHOW", Object.values(json.pin))
+    }
+  }
+}
+
+const handleConsumerMessage = async (message: string): Promise<void> => {
+  if (ls.step === "NEGOTIATION") {
+    const pin = await handleSessionKey(message)
+    if (pin) {
+      console.log("PIN to SHOW", Array.from(pin))
+      await sendUserChallenge(pin)
+    }
+  }
+}
+
+// 🔗 Device Linking Steps 
+
+/**
+ * CONSUMER: BROADCAST
+ * 
+ * This is the CONSUMER first step (after opening the channel) where we 
+ * broadcast a temporary public key. This key is then published on channel as 
+ * the throwaway DID key.
+ */
+export const sendTemporaryExchangeKey = async (): Promise<void> => {
+  const cfg = config.normalize()
+
+  const { rsaSize, hashAlg } = cfg
+  ls.temporaryRsaPair = await rsa.makeKeypair(rsaSize, hashAlg, KeyUse.Exchange)
+  const pubKey = await rsa.getPublicKey(ls.temporaryRsaPair)
+  const temporaryDID = did.publicKeyToDid(pubKey, did.KeyType.RSA)
+  await publishOnChannel(temporaryDID)
+  nextStep()
+}
+
+/**
+ * state: PRODUCER: BROADCAST
+ * 
+ * This should be called by the PRODUCER upon receiving the throwaway DID key.
+ * 
+ * @param didThrowaway 
+ */
+export const sendSessionKey = async (didThrowaway: string): Promise<void> => {
+  ls.sessionKey = await aes.makeKey({ alg: SymmAlg.AES_GCM, length: 256 })
+
+  const sessionKey = await aes.exportKey(ls.sessionKey)
+
+  const { publicKey } = did.didToPublicKey(didThrowaway)
+  const publicCryptoKey = await rsa.importPublicKey(publicKey, HashAlg.SHA_256, KeyUse.Exchange)
+
+  // Note: rsa.encrypt expects a B16 string
+  const rawSessionKey = utils.arrBufToStr(utils.base64ToArrBuf(sessionKey), CharSize.B16)
+  const encryptedSessionKey = await rsa.encrypt(rawSessionKey, publicCryptoKey)
+
+  const u = await ucan.build({
+    issuer: await did.ucan(),
+    audience: didThrowaway,
+    lifetimeInSeconds: 60 * 5, // 5 minutes
+    facts: [{ sessionKey }],
+    potency: null
+  })
+
+  const iv = utils.randomBuf(16)
+  const msg = await aes.encrypt(ucan.encode(u), ls.sessionKey, { iv, alg: SymmAlg.AES_GCM })
+
+  await publishOnChannel(
+    JSON.stringify({
+      iv: utils.arrBufToBase64(iv),
+      msg,
+      sessionKey: utils.arrBufToBase64(encryptedSessionKey)
+    })
```

Comment:
Reminder for @bgins - this is where the auth-lobby and whitepaper have diverged (sending the UCAN encrypted by the AES key and the AES key separately encrypted by the RSA key).

Reply:
I discussed this with @expede this morning, and this implementation is correct. While it would be nice to encrypt with the temporary RSA key, we can't fit enough data in with RSA. 

Instead, we encrypt the AES key with the temporary RSA key and the UCAN with the AES key. On the other end, the AES key is decrypted and then used to decrypt the UCAN.